File: bintree_java/src/binarytree/AvlTree.java
package binarytree;


import java.util.List;



public class AvlTree< TKey extends Comparable<? super TKey> >
             extends BinSearchTree< TKey, AvlNode<TKey> >
{


//////////////////////////////////////////////////////////////
//                        FIELDS
//////////////////////////////////////////////////////////////



//////////////////////////////////////////////////////////////
//                        CONSTRUCTORS
//////////////////////////////////////////////////////////////



    public AvlTree() {
    }



    public AvlTree(CandidateRemoval canddRM) {
        super(canddRM);
    }



    public AvlTree(List<TKey> lst) {
        super(lst);
    }



    public AvlTree(List<TKey> lst, CandidateRemoval canddRM) {
        super(lst, canddRM);
    }



//////////////////////////////////////////////////////////////
//                        METHODS (PUBLIC)
//////////////////////////////////////////////////////////////



    @Override
    public int height() {
        return (null == this.root) ? 0 : this.root.height();
    }



    @Override
    public AvlTree<TKey> clone() {
        var theClone = new AvlTree<TKey>();

        if (null != this.root) {
            theClone.root = this.root.cloneBranch();
        }

        theClone.traversal = this.traversal;
        theClone._count = this._count;
        theClone.optionCanddRM = this.optionCanddRM;

        return theClone;
    }



//////////////////////////////////////////////////////////////
//                        METHODS (PROTECTED)
//////////////////////////////////////////////////////////////



    @Override
    protected AvlNode<TKey> _insert(AvlNode<TKey> node, TKey key) {
        node = super._insert(node, key);
        node = adjustBalance(node);
        return node;
    }



    @Override
    protected AvlNode<TKey> _remove(AvlNode<TKey> node, TKey key) {
        node = super._remove(node, key);

        if (null != node) {
            node = adjustBalance(node);
        }

        return node;
    }



    protected AvlNode<TKey> adjustBalance(AvlNode<TKey> node) {
        // STEP 1. Update the height of the node
        node.updateHeight();

        // STEP 2. Get the balance factor
        int balance = node.balance();
        int balanceLe = node.balanceLeft();
        int balanceRi = node.balanceRight();

        // STEP 3. Process if the node is unbalanced ==> 4 cases

        // Case 1: left-left
        if (  balance > 1 && balanceLe >= 0  )
            return rotateRight(node);

        // Case 2: right-right
        if (  balance < -1 && balanceRi <= 0  )
            return rotateLeft(node);

        // Case 3: left-right
        if (  balance > 1 && balanceLe < 0  ) {
            node.left = rotateLeft(node.left);
            return rotateRight(node);
        }

        // Case 4: right-left
        if (  balance < -1 && balanceRi > 0  ) {
            node.right = rotateRight(node.right);
            return rotateLeft(node);
        }

        return node;
    }



    protected AvlNode<TKey> rotateLeft(AvlNode<TKey> node) {
        /*
            node
            /  \
           ..   V
               / \
              a   b
        */
        var V = node.right;
        var a = V.left;

        V.left = node;
        node.right = a;

        node.updateHeight();
        V.updateHeight();

        return V;
    }



    protected AvlNode<TKey> rotateRight(AvlNode<TKey> node) {
        /*
            node
            /  \
           V   ..
          / \
         a   b
        */
        var V = node.left;
        var b = V.right;

        V.right = node;
        node.left = b;

        node.updateHeight();
        V.updateHeight();

        return V;
    }



    @Override
    protected AvlNode<TKey> createNode(TKey key) {
        return new AvlNode<TKey>(key);
    }



    @Override
    protected void buildTreeFromSortedListNodeFunc(AvlNode<TKey> node) {
        node.updateHeight();
    }
}

File: bintree_csharp/binarytree/AvlTree.cs
using System;
using System.Collections.Generic;
using my.extensions;


namespace my.binarytree
{
    class AvlTree<TKey> : BinSearchTree<TKey, AvlNode<TKey>>
        where TKey : IComparable
    {

        //////////////////////////////////////////////////////////////
        //                        CONSTRUCTORS
        //////////////////////////////////////////////////////////////



        public AvlTree()
        {
        }



        public AvlTree(CandidateRemoval canddRM) : base(canddRM)
        {
        }



        public AvlTree(List<TKey> lst) : base(lst)
        {
        }



        public AvlTree(List<TKey> lst, CandidateRemoval canddRM) : base(lst, canddRM)
        {
        }



        //////////////////////////////////////////////////////////////
        //                        PROPERTIES
        //////////////////////////////////////////////////////////////



        public override int Height => (this.Root is null) ? 0 : this.Root.Height;



        //////////////////////////////////////////////////////////////
        //                        METHODS (PUBLIC)
        //////////////////////////////////////////////////////////////



        public override AvlTree<TKey> Clone() => this.DeepCopy();



        //////////////////////////////////////////////////////////////
        //                        METHODS (PROTECTED)
        //////////////////////////////////////////////////////////////



        protected override AvlNode<TKey> _Insert(AvlNode<TKey> node, TKey key)
        {
            node = base._Insert(node, key);
            node = AdjustBalance(node);
            return node;
        }



        protected override AvlNode<TKey> _Remove(AvlNode<TKey> node, TKey key)
        {
            node = base._Remove(node, key);

            if (node is not null)
            {
                node = AdjustBalance(node);
            }

            return node;
        }



        protected AvlNode<TKey> AdjustBalance(AvlNode<TKey> node)
        {
            // STEP 1. Update the height of the node
            node.UpdateHeight();

            // STEP 2. Get the balance factor
            int balance = node.Balance;
            int balanceLe = node.BalanceLeft;
            int balanceRi = node.BalanceRight;

            // STEP 3. Process if the node is unbalanced ==> 4 cases

            // Case 1: left-left
            if (balance > 1 && balanceLe >= 0)
                return RotateRight(node);

            // Case 2: right-right
            if (balance < -1 && balanceRi <= 0)
                return RotateLeft(node);

            // Case 3: left-right
            if (balance > 1 && balanceLe < 0)
            {
                node.Left = RotateLeft(node.Left);
                return RotateRight(node);
            }

            // Case 4: right-left
            if (balance < -1 && balanceRi > 0)
            {
                node.Right = RotateRight(node.Right);
                return RotateLeft(node);
            }

            return node;
        }



        protected AvlNode<TKey> RotateLeft(AvlNode<TKey> node)
        {
            /*
                node
                /  \
               ..   V
                   / \
                  a   b
            */
            var V = node.Right;
            var a = V.Left;

            V.Left = node;
            node.Right = a;

            node.UpdateHeight();
            V.UpdateHeight();

            return V;
        }



        protected AvlNode<TKey> RotateRight(AvlNode<TKey> node)
        {
            /*
                node
                /  \
               V   ..
              / \
             a   b
            */
            var V = node.Left;
            var b = V.Right;

            V.Right = node;
            node.Left = b;

            node.UpdateHeight();
            V.UpdateHeight();

            return V;
        }



        protected override void BuildTreeFromSortedListNodeFunc(AvlNode<TKey> node) => node.UpdateHeight();
    }
}
